### PR TITLE
pgwire: remove readTimeoutConn in favor of a channel

### LINKED ISSUE #25585: pgwire: replace deadline and connection polling with separate goroutine
From https://github.com/cockroachdb/cockroach/pull/25328#issuecomment-386760916:

>Rather than setting a deadline on the read and periodically polling whether the session is cancelled, we could fire up another goroutine which blocks on ctx.Done() and then closes the connection. I recall there was some reason this approach wasn't taken when this code was written, but taking another look it isn't clear to me why it wouldn't work. This would be a bit more involved of a change, so I'm fine with the current PR as a stop-gap solution.

The reason we use a deadline approach in the first place is because:
>even when the context is canceled, you still might need to use the connection to send back an error (e.g. sending an AdminShutdownError when draining).

However, this can be worked around by half closing the connection. That is, we can close the connection for reading but still write to it. See [TCPConn.CloseRead](https://golang.org/pkg/net/#TCPConn.CloseRead).

Jira issue: CRDB-5704

### PR DESCRIPTION
**Pull request by rafiss**
Rather than using a connection that polls for the context being done every second, we now spin up an additional goroutine that blocks until the connection context is done, or the drain signal was received.

I wrote a simple benchmark of the idle connection loop to generate CPU profiles.

With the old readTimeoutConn:
image

With the new goroutine approach:
image

There's definitely less noise and overhead.

fixes https://github.com/cockroachdb/cockroach/issues/25585
Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Thanks for looking into this! The change overall seems good to me.

I'm not too familiar with this code, so I want to double check my understanding. IIUC currently for each connection we have two goroutines:
- the "reader" goroutine that continuously loops in `serveImpl` and see what the client is requesting us to do
- and the "processor" goroutine for the connExecutor that executes commands added into `stmtBuf` by the "reader" goroutine.

This PR introduces a third "watchdog" goroutine that monitors the state for cancellation and draining. In other words, we make a trade-off between 1) periodically forcing `net.Conn` to wake up in the "reader" due to a read deadline to see whether the "reader" should quit and 2) a separate goroutine that only monitors things with no read deadline set for the "reader". In practice, this "watchdog" goroutine should be idle most of the time, so it should remove some system calls. Does this sound right?

File: pkg/sql/pgwire/server.go [hunk 1]
```diff
@@ -945,6 +942,35 @@ func (s *Server) serveImpl(
...
+		var tcpConn *net.TCPConn
+		switch c := c.conn.(type) {
+		case *net.TCPConn:
+			tcpConn = c
+		case *tls.Conn:
+			underConn := c.NetConn()
+			tcpConn, _ = underConn.(*net.TCPConn)
+		}
+		if tcpConn == nil {
+			_ = c.conn.Close()
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/pgwire/server.go:968]: hm good point. it seems safer to only close the read side.

this matches the previous behavior, where if the context was cancelled, then the readTimeoutConn would start returning errors when `Read` is called: https://github.com/cockroachdb/cockroach/blob/ceb19c5b75cf4aa02c68a669122349a3e9aca33d/pkg/sql/pgwire/server.go#L1039

that should cause the main loop of the reader goroutine to exit here: https://github.com/cockroachdb/cockroach/blob/ceb19c5b75cf4aa02c68a669122349a3e9aca33d/pkg/sql/pgwire/server.go#L1216-L1225

**Review by rafiss — COMMENTED** [review]
> This PR introduces a third "watchdog" goroutine that monitors the state for cancellation and draining. In other words, we make a trade-off between 1) periodically forcing net.Conn to wake up in the "reader" due to a read deadline to see whether the "reader" should quit and 2) a separate goroutine that only monitors things with no read deadline set for the "reader". In practice, this "watchdog" goroutine should be idle most of the time, so it should remove some system calls. Does this sound right?

yes sounds right. it's worth noting that a goroutine that does nothing apart from wait on a channel is very efficient in go. the scheduler should never try to run it.

File: pkg/sql/pgwire/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/pgwire/server.go:968]: although, thinking about it more, the risk is that if the type switch no longer works as expected (like for example, there's a new implementation of net.Conn to deal with later), then not having a default case would make the connection hang forever.

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/pgwire/server.go` line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124373#gh-1605895348:-NyM96ArEhgM8YVuM9ZG:bhs0ual) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f528eda54761c0956113c558fa96f7c60bfd12/pkg/sql/pgwire/server.go#L968)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

although, thinking about it more, the risk is that if the type switch no longer works as expected (like for example, there's a new implementation of net.Conn to deal with later), then not having a default case would make the connection hang forever.

maybe it's safer to keep the Close as a fallback? WDYT?
</blockquote></details>

Sounds reasonable to me.

I traced where `Close` is called currently, and I think it's a few layers up in the `defer` in in the async task in `netutil.TCPServer.ServeWith`, so the comment (that "serveImpl always closes the network connection before returning") is misleading before your patch but will actually become closer to reality after the patch. It might be worth adjusting the comment.

**Review by yuzefovich — COMMENTED** [review]
Looks like the updates haven't been pushed yet?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373#-:-NyMEcK6ABx9LjCKaw7Q:b58epck)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/pgwire/server.go` line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124373#gh-1605895348:-NyMEzZZ-4EddJRtyPPR:b-50d36w) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f528eda54761c0956113c558fa96f7c60bfd12/pkg/sql/pgwire/server.go#L968)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Sounds reasonable to me.

I traced where `Close` is called currently, and I think it's a few layers up in the `defer` in in the async task in `netutil.TCPServer.ServeWith`, so the comment (that "serveImpl always closes the network connection before returning") is misleading before your patch but will actually become closer to reality after the patch. It might be worth adjusting the comment.
</blockquote></details>

I was about to push the change, then realized that another fallback option is to call `conn.SetReadDeadline(now())`, which might be less disruptive. I'll do that instead.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373#-:-NyMH1Yx0n5f7YhDMPpO:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/pgwire/server.go` line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124373#gh-1605895348:-NyMGgiDAlu4npGk_njU:b-5cxqux) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f528eda54761c0956113c558fa96f7c60bfd12/pkg/sql/pgwire/server.go#L968)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I was about to push the change, then realized that another fallback option is to call `conn.SetReadDeadline(now())`, which might be less disruptive. I'll do that instead.
</blockquote></details>

Nice, I like it.

**Comment by rafiss** [comment]
tftr!

bors r=yuzefovich

**Comment by rickystewart** [comment]
bors r=yuzefovich

**Pull request merged**